cdap-ui/app/cdap/components/DataPrep/DataPrepServiceControl/index.js

```diff
@@ -0,0 +1,163 @@
+import MyDataPrepApi from 'api/dataprep';
+import {MyArtifactApi} from 'api/artifact';
+import find from 'lodash/find';
+import NamespaceStore from 'services/NamespaceStore';
+
+export default class DataPrepServiceControl extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      loading: false,
+      error: null,
+      extendedMessage: null
+    };
+
+    this.enableService = this.enableService.bind(this);
+  }
+
+  componentWillUnmount() {
+    if (this.servicePoll && this.servicePoll.dispose) {
+      this.servicePoll.dispose();
+    }
+  }
+
+  enableService() {
+    this.setState({loading: true});
+
+    /**
+     *  1. Get Wrangler Service App
+     *  2. If not found, create app
+     *  3. Start Wrangler Service
+     *  4. Poll until service starts, then reload page
+     **/
+
+    let namespace = NamespaceStore.getState().selectedNamespace;
+
+    MyDataPrepApi.getApp({ namespace })
+      .subscribe(() => {
+        // Wrangler app already exist
+        // Just start service
+        this.startService();
+      }, () => {
+        // App does not exist
+        // Go to create app
+        this.createApp();
+      });
+  }
+
+  createApp() {
+    let namespace = NamespaceStore.getState().selectedNamespace;
+
+    MyArtifactApi.list({ namespace })
+      .subscribe((res) => {
+        let artifact = find(res, { 'name': 'wrangler-service' });
+
+        MyDataPrepApi.createApp({ namespace }, { artifact })
+          .subscribe(() => {
+            this.startService();
+          }, (err) => {
+            this.setState({
+              error: 'Failed to start service',
+              extendedMessage: err.data || err,
+              loading: false
+            });
+          });
+      });
+  }
+
+  startService() {
+    let namespace = NamespaceStore.getState().selectedNamespace;
+
+    MyDataPrepApi.startService({ namespace })
+      .subscribe(() => {
+        this.pollServiceStatus();
+      }, (err) => {
+        this.setState({
+          error: 'Failed to start service',
+          extendedMessage: err.data || err,
+          loading: false
+        });
+      });
+  }
+
+  pollServiceStatus() {
+    let namespace = NamespaceStore.getState().selectedNamespace;
+
+    this.servicePoll = MyDataPrepApi.pollServiceStatus({ namespace })
+      .subscribe((res) => {
+        if (res.status === 'RUNNING') {
+          window.onbeforeunload = null;
+          window.location.reload();
+        }
+      }, (err) => {
+        this.setState({
+          error: 'Failed to get service status',
+          extendedMessage: err.data || err,
+          loading: false
+        });
+      });
+  }
+
+  renderError() {
+    if (!this.state.error) { return null; }
+
+    return (
+      <div>
+        <h5 className="text-danger text-xs-center">
+          {this.state.error}
+        </h5>
+        <p className="text-dangertext-xs-center">
+          {this.state.extendedMessage}
```

Comment:
Might want to check if `extendedMessage` is an object. It might be `response & statusCode` object.,